core/dbt/include/global_project/macros/materializations/common/merge.sql

```diff
@@ -37,8 +38,8 @@
 
     {% if unique_key %}
     when matched then update set
-        {% for column in dest_columns -%}
-            {{ adapter.quote(column.name) }} = DBT_INTERNAL_SOURCE.{{ adapter.quote(column.name) }}
+        {% for column_name in update_columns -%}
+            {{ adapter.quote(column_name) }} = DBT_INTERNAL_SOURCE.{{ adapter.quote(column_name) }}
```

Comment:
And then this line as well:
```suggestion
            {{ column_name }} = DBT_INTERNAL_SOURCE.{{ column_name }}
```
If using `dest_columns`, they're already quoted. If using `merge_update_columns`, leave it up to the user to case-and-quote if the column name requires quoting (e.g. mixed casing or special characters).